src/Core/Category/NameBuilder/CategoryDisplayNameBuilder.php

```diff
@@ -0,0 +1,265 @@
+            return Cache::retrieve($cacheKey);
+        }
+
+        $duplicateCategoriesBreadcrumbs = $this->fetchBreadcrumbs($categoryIds, $languageId);
+        Cache::store($cacheKey, $duplicateCategoriesBreadcrumbs);
+
+        return $duplicateCategoriesBreadcrumbs;
+    }
+
+    /**
+     * @param CategoryId[] $categoryIds
+     * @param LanguageId $languageId
+     *
+     * @return array<int, string[]>
+     */
+    private function fetchBreadcrumbs(array $categoryIds, LanguageId $languageId): array
+    {
+        $duplicateCategoriesBreadcrumbs = [];
+        foreach ($categoryIds as $categoryId) {
+            $duplicateCategoriesBreadcrumbs[$categoryId->getValue()] = $this->categoryRepository->getBreadcrumbParts(
+                $categoryId,
+                $languageId
+            );
+        }
+
+        return $duplicateCategoriesBreadcrumbs;
+    }
+
+    /**
+     * Builds a breadcrumb which consists of minimum needed parents to be unique, depending on other categories breadcrumbs.
+     * If breadcrumb is still not unique after showing all of it, then we append category id.
+     *
+     * E.g.:
+     *      For categories: Home > Clothes, Home > Clothes
+     *          breadcrumbs would be: Home > Clothes (#id1), Home > Clothes (#id2)
+     *
+     *      For: Home > Clothes, Home > Clothes > Clothes:
+     *          breadcrumbs would be: Home > Clothes, Clothes > Clothes
+     *
+     * @param CategoryId $categoryId
+     * @param array<int, string[]> $duplicateCategoriesBreadcrumbs
+     *
+     * @return string
+     */
+    private function buildBreadcrumb(CategoryId $categoryId, array $duplicateCategoriesBreadcrumbs): string
+    {
+        $breadcrumbParts = $duplicateCategoriesBreadcrumbs[$categoryId->getValue()];
+        unset($duplicateCategoriesBreadcrumbs[$categoryId->getValue()]);
+
+        $maxSteps = count($breadcrumbParts);
+        $breadcrumb = $breadcrumbParts[0];
+        $duplicatedFound = true;
+
+        if ($maxSteps > 1) {
+            $duplicatedFound = false;
+
+            for ($step = 2; $step <= $maxSteps; ++$step) {
+                $breadcrumb = $this->extractBreadcrumbFromParts($breadcrumbParts, $step);
+                $duplicatedFound = false;
+                foreach ($duplicateCategoriesBreadcrumbs as $otherCategoryBreadcrumbs) {
+                    $otherBreadcrumb = $this->extractBreadcrumbFromParts($otherCategoryBreadcrumbs, $step);
+                    if ($otherBreadcrumb === $breadcrumb) {
+                        $duplicatedFound = true;
+                        break;
+                    }
+                }
+
+                if (!$duplicatedFound) {
+                    break;
+                }
+            }
+        }
+
+        if ($duplicatedFound) {
+            // if breadcrumbs are still duplicated append category id
+            $breadcrumb = sprintf('%s (#%d)', $breadcrumb, $categoryId->getValue());
+        }
```

Comment:
IMO the category ID should be present in all ctg names, duplicateds or not.

Reply:
That is for @MatShir to decide. But i don't think its very useful (maybe for developers yes, but probably not merchants)